crates/sui-graphql-rpc/src/types/object.rs

```diff
@@ -1233,59 +1248,105 @@ impl Loader<HistoricalKey> for Db {
+            .map_err(|e| Error::Internal(format!("Failed to fetch objects: {e}")))?;
+        let native_objects: Vec<StoredFullHistoryObject> = native_objects_result
             .map_err(|e| Error::Internal(format!("Failed to fetch objects: {e}")))?;
 
-        let mut id_version_to_stored = BTreeMap::new();
-        for stored in objects {
-            let key = (addr(&stored.object_id)?, stored.object_version as u64);
-            id_version_to_stored.insert(key, stored);
+        let mut id_version_to_indexed = BTreeMap::new();
+        for indexed in indexed_objects {
+            let key = (addr(&indexed.object_id)?, indexed.object_version as u64);
+            id_version_to_indexed.insert(key, indexed);
+        }
+        let mut id_version_to_native = BTreeMap::new();
+        for native in native_objects {
+            let key = (addr(&native.object_id)?, native.object_version as u64);
+            id_version_to_native.insert(key, native);
         }
 
         let mut result = HashMap::new();
         for key in keys {
-            let Some(stored) = id_version_to_stored.get(&(key.id, key.version)) else {
-                continue;
-            };
+            let indexed_opt = id_version_to_indexed.get(&(key.id, key.version));
 
             // Filter by key's checkpoint viewed at here. Doing this in memory because it should be
             // quite rare that this query actually filters something, but encoding it in SQL is
             // complicated.
-            if key.checkpoint_viewed_at < stored.checkpoint_sequence_number as u64 {
-                continue;
-            }
+            let object = if let Some(indexed) = indexed_opt {
+                if key.checkpoint_viewed_at < indexed.checkpoint_sequence_number as u64 {
+                    continue;
+                }
+                Object::try_from_stored_history_object(
+                    indexed.clone(),
+                    key.checkpoint_viewed_at,
+                    // This conversion will use the object's own version as the `Object::root_version`.
+                    None,
+                )?
+            } else {
+                let Some(native) = id_version_to_native.get(&(key.id, key.version)) else {
+                    continue;
+                };
+                if let Some(serialized) = &native.serialized_object {
+                    let object = bcs::from_bytes(serialized).map_err(|_| {
+                        Error::Internal(format!("Failed to deserialize object at {:?}", key))
+                    })?;
+                    Object::from_native(
+                        key.id,
+                        object,
+                        key.checkpoint_viewed_at,
+                        // This conversion will use the object's own version as the `Object::root_version`.
+                        None,
+                    )
+                } else {
+                    Object::new_wrapped_or_deleted(key.id, key.version, key.checkpoint_viewed_at)
```

Comment:
IIUC, this is a bit strange, it means that if I request some totally made up ObjectID and/or version, I will get a response that says that that object is wrapped or deleted, which doesn't seem right.

I'm onboard with the idea that we shouldn't distinguish between an object that is wrapped/deleted/pruned/never existed, but I think in practice that has to mean that they all return no response, rather than this response.

Reply:
In this case, we know for sure it is a wrapped/deleted object, because we had an entry from full_objects_history table, but the serialized column is null.